cranelift/isle/isle/src/trie_again.rs

```diff
@@ -318,66 +318,130 @@ impl RuleSetBuilder {
         }
     }
 
-    // If a binding site is constrained and also required to be equal to another binding site, then
-    // copy the constraint and push the equality inside it.
+    /// Establish the invariant that a binding site can have a concrete constraint or a nonlinear
+    /// constraint, but not both. This is useful because overlap checking is most effective on
+    /// concrete constraints, and also because it exposes more rule structure for codegen.
+    ///
+    /// If a binding site is constrained and also required to be equal to another binding site, then
+    /// copy the constraint and push the equality inside it. For example:
+    /// - `(term x @ 2 x)` is rewritten to `(term 2 2)`
+    /// - `(term x @ (T.A _) x)` is rewritten to `(term (T.A y) (T.A y))`
```

Comment:
Hmm -- now that I think about it more, I'm not sure this is always a valid transform as a general IR normalization/canonicalization (as opposed to something useful just for overlap-checking). Specifically, I think it loses information -- consider

```lisp
(term x @ (T.A y _) x)  --> (term (T.A y _) (T.A y _))
```

if we then match against the input `(term (T.A 1 2) (T.A 1 3))`, the original form will not match, but the rewritten form will match (incorrectly, wrt the original pattern).

I suppose this is fine for overlap checking -- the rule always "expands" (matches more) under this transform, I think, since we can lose equality constraints -- but I think it is only exact (my inner thermodynamicist is going to invent the word "isoconstraining" for this) if the constraint on the constrained-and-equals site *fully captures the value* (if a variant, all fields have patterns that, recursively, fully capture the value, or are just a variable binding).

So if we hope to use this IR as a basis for compilation, I think we need to avoid this transform, though I suppose we will want to find a way to get the same effect for overlap checking, unless we have a more direct way of handling the equality constraints. (It feels like maybe there should be a more direct way to handle the equality constraints, but I don't have any good ideas right now, sorry!)

Reply:
Ah, and now that I write all that, I see the `y` term in the above, and I've looked at the code again and am understanding the bit about handling all bindings within the variant -- I think I've convinced myself of the need for what you are actually doing already. So, nevermind all of the above :-)